### DISCUSSION TITLE
Is there a way to handle concatenating dataset if data_vars differ?

### DISCUSSION THREAD

**Discussion by ahuang11** [pydata/xarray#4586] [Q&A] [answered]
```
import xarray as xr
import pandas as pd
ds1 = xr.tutorial.open_dataset('air_temperature')
ds2 = ds1.copy()
ds2['airx2'] = ds1['air'] * 2
ds2['time'] = ds2['time'] + pd.Timedelta(days=1000)
ds2 = ds2.drop('air')
ds3 = xr.concat([ds1, ds2], 'time')
```
`ValueError: 'air' is not present in all datasets.`

Is there a way to get around this? I tried setting data_vars='different' but still errors.

**Comment by ahuang11** [comment]
I guess combine_by_coords seems to work, but I wonder if concat can have a keyword to handle this automatically?
`ds3 = xr.combine_by_coords([ds1, ds2])`

**Comment by keewis** [comment, marked as answer]
Unfortunately, the documentation of the `data_vars` parameter is a bit unclear, so we need to look at the code to know its purpose. The variables to be concatenated are calculated in `_calc_concat_over` (which could use some refactoring, by the way): https://github.com/pydata/xarray/blob/dd9fe2a8a414ddefa3b04b934163c9ccc628c5c7/xarray/core/concat.py#L220-L333
If I read that correctly, all variables with `dim` are added to the set of variables to be concatenated. Every value other than `data_vars="minimal"` extends that with variables that don't have that dimension.

As a summary, there are two documentation bugs: we don't explicitly state that concatenated variables have to be present in all objects (we currently only learn about that from the error message), and the documentation of `data_vars` could use some polishing.

I guess the conclusion is that `xr.concat` is just not the right tool for your use case, and that it would probably be better to use `xr.merge` or a related function like `xr.combine_by_coords`?

**Reply by ahuang11** [reply]
Thanks! I didn't know xr.merge could concatenate coordinates automatically too.

**Comment by spencerkclark** [comment]
This is indeed often awkward.  There's been some work and discussion around this before, but nothing yet merged; see https://github.com/pydata/xarray/pull/3545 and https://github.com/pydata/xarray/pull/508.
